cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh

```diff
@@ -0,0 +1,301 @@
+  {
+    // compute offsets from num_geoms_per_pair
+    thrust::uninitialized_fill_n(rmm::exec_policy(stream), offsets->begin(), offsets->size(), 0);
+    thrust::inclusive_scan(rmm::exec_policy(stream),
+                           num_geoms_per_pair.begin(),
+                           num_geoms_per_pair.end(),
+                           thrust::next(offsets->begin()));
+  }
+
+  /// Return range to offset array
+  auto offset_range() { return range{offsets->begin(), offsets->end()}; }
+
+  /// Return range to geometry array
+  auto geom_range() { return range{geoms->begin(), geoms->end()}; }
+
+  /// Return id_range structure to id arrays
+  auto get_id_ranges()
+  {
+    return id_ranges{range(lhs_linestring_ids->begin(), lhs_linestring_ids->end()),
+                     range(lhs_segment_ids->begin(), lhs_segment_ids->end()),
+                     range(rhs_linestring_ids->begin(), rhs_linestring_ids->end()),
+                     range(rhs_segment_ids->begin(), rhs_segment_ids->end())};
+  }
+
+  /// Return the number of pairs in the intermediates
+  auto size() { return offsets->size() - 1; }
+};
+
+/**
+ * @brief Kernel to compute intersection and store the result and look-back indices to outputs.
+ *
+ * Each thread operates on one segment in `multilinestrings1`, iterates over all segments in the
+ * other multilinestring.
+ */
+template <typename MultiLinestringRange1,
+          typename MultiLinestringRange2,
+          typename TempIt1,
+          typename TempIt2,
+          typename Offsets1,
+          typename Offsets2,
+          typename Offsets3,
+          typename IdRanges,
+          typename OutputIt1,
+          typename OutputIt2>
+void __global__ pairwise_linestring_intersection_simple(MultiLinestringRange1 multilinestrings1,
+                                                        MultiLinestringRange2 multilinestrings2,
+                                                        TempIt1 n_points_stored,
+                                                        TempIt2 n_segments_stored,
+                                                        Offsets1 num_points_offsets_first,
+                                                        Offsets2 num_segments_offsets_first,
+                                                        Offsets3 num_points_per_pair_first,
+                                                        IdRanges point_ids_range,
+                                                        IdRanges segment_ids_range,
+                                                        OutputIt1 points_first,
+                                                        OutputIt2 segments_first)
+{
+  using T       = typename MultiLinestringRange1::element_t;
+  using types_t = uint8_t;
+  using count_t = iterator_value_type<Offsets1>;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
```

Comment:
Isn't there a utility function for this?

Reply:
Ha! Mark asked the similar question here: https://github.com/rapidsai/cuspatial/pull/813#discussion_r1034163231. Short answer is no. But we can create a short hand for it with `counting_iterator` and `range`. Want me to do so?

Reply:
Or introduce https://github.com/harrism/ranger to the code base.

Reply:
I think it's worth making an issue to track this. But probably outside of the scope of this PR.